src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,175 @@
-        return false;
+        if (list.size() == 0) {
+            return false;
+        }
+        for (T element : list) {
+            linkLast(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).value;
+
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> x = node(index);
+        T oldValue = x.value;
+        x.value = value;
+        return oldValue;
+    }
+
+    Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            Node<T> node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        } else {
+            Node<T> node = last;
+            for (int i = size - 1; i > index; i--) {
+                node = node.prev;
+            }
+            return node;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
-    public boolean remove(T t) {
+    public boolean remove(T value) {
+        if (value == null) {
+            for (Node<T> node = first; node != null; node = node.next) {
+                if (node.value == null) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> node = first; node != null; node = node.next) {
+                if (value.equals(node.value)) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.value = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    void linkLast(T e) {
+        Node<T> l = last;
+        Node<T> newNode = new Node<>(l, e, null);
+        last = newNode;
+
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
+
+    }
+
+    void linkBefore(T e, Node<T> success) {
+        Node<T> pred = success.prev;
+        Node<T> newNode = new Node<>(pred, e, success);
+        success.prev = newNode;
+
+        if (pred == null) {
+            first = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
+    }
+
+    public T unlink(Node<T> node) {
+        final T element = node.value;
```

Comment:
final-? Why do we use it here?

Reply:
Maven says that
Distance between variable 'element' declaration and its first usage is 5, but allowed 3.  Consider making that variab
le final if you still need to store its value in advance

Reply:
done